### ISSUE TITLE
uninformative error location for termination error

### ISSUE THREAD

**Issue by rolph-recto** [ucsd-progsys/liquidhaskell#424]
Prof. Jhala,
This is a follow up to our meeting last Tuesday. I inserted a bug in the recursive call to `perms`, like so:

    module permutation () where

    {-@ permutations :: ts:[a] -> [[a]] / [(len ts), 1, 0] @-}
    permutations            :: [a] -> [[a]]
    permutations xs0        =  xs0 : perms xs0 []

    {-@ perms :: ts:[a] -> is:[a] -> [[a]] / [((len ts)+(len is)), 0, (len ts)] @-}
    perms :: [a] -> [a] -> [[a]]
    perms []     _  = []
    -- permutation call in body supposed to have arg (is) not (t:is)
    perms (t:ts) is = foldr interleave (perms ts (t:is)) (permutations (t:is))
          where interleave    xs     r = let (_,zs) = interleave' id xs r in zs
                interleave' _ []     r = (ts, r)
                interleave' f (y:ys) r = let (us,zs) = interleave' (f . (y:)) ys r
                                         in  (y:us, f (t:y:us) : zs)

passing this to `liquid` gives the error message:

    **** UNSAFE ********************************************************************
    Permutation.hs:9:1-5: Error: Termination Check

which maps back to the function name of `perms` in the base case, *not* the recursive case. I don't think we agreed on the correct source location to implicate, but you commented that the termination error should report back a different source location since this one is uninformative.

**Comment by nikivazou** [comment]
Is the suggestion here that the error shows up at the recursive call? I think this is feasible!

**Comment by ranjitjhala** [comment]
Yes, it should go to the recursive call (where the actual "decreasing call"
happens).

​
